Fast mode task config + vendor config task

tobonex opened the pull request:
First PR here, open to heavy criticism.

Adding vendor config set and fmt(fast mode task) config as first steps to fmt implementation. Fixes IPC error on fmt config due to no fmt implementation.

sofci commented:
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

gkbldcig commented:
Can one of the admins verify this patch?

btian1 reviewed (commented):
Please fix CI issues first

lgirdwood commented on src/include/sof/audio/kpb.h:
```diff
@@ -159,18 +163,89 @@ struct history_data {
+	struct {
+		uint16_t module_id;
+		uint16_t instance_id;
+	}		    dev_ids[1];
```
indentation

lgirdwood reviewed (commented):
Needs more decription in the commit messages to explain feature and changes.

tmleman commented on src/include/ipc4/module.h:
```diff
@@ -203,6 +206,35 @@ struct ipc4_module_bind_unbind {
+	uint32_t param_size;
+	//! value (padded w/ zeroes to align to dword)
+	uint32_t param_data[1];
+};
```
We have struct that represent TLV, use it instead.

https://github.com/thesofproject/sof/blob/main/src/include/sof/tlv.h

btian1 commented on src/audio/kpb.c:
```diff
@@ -2384,9 +2384,17 @@ static int kpb_set_large_config(struct comp_dev *dev, uint32_t param_id,
+	/* We can use extended param id for both extended
+	 * and standard param id
+	 */
+	const struct byte_array_simple *array = (struct byte_array_simple *)data;
```
redundant line 2390?

btian1 commented on src/ipc/ipc4/handler.c:
```diff
@@ -828,6 +828,100 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
+	/* else, !(init_block && final_block) */
+	const struct ipc4_mod_conf_param *tlv = (struct ipc4_mod_conf_param *)data;
+	struct byte_array_simple array = {
+			(uint8_t *)&data, data_off_size };
```
should no need use two lines.

lyakh commented on src/include/sof/audio/kpb.h:
```diff
@@ -159,18 +163,89 @@ struct history_data {
+
+struct fast_mode_task {
+	/*! Array of pointers to all module lists to be processed. */
+	struct device_list *device_list_[FAST_MODE_TASK_MAX_LIST_COUNT];
```
can we avoid ending names with an underscore?

lyakh commented on src/audio/kpb.c:
```diff
@@ -2386,14 +2591,42 @@ static int kpb_set_large_config(struct comp_dev *dev, uint32_t param_id,
+
+		if (nr_of_modules != 0) {
+			cfg = (struct kpb_task_params *)array->data;
+			if (ret  < 0)
```
wrong test

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,190 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+	assert(cfg && pin < KPB_MAX_SINK_CNT && pin != REALTIME_PIN_ID &&
+	       cfg->dev_ids > 0);
+
+	int ret = 0;
```
superfluous initialisation

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,190 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+				 size_t pin)
+{
+	assert(cfg && pin < KPB_MAX_SINK_CNT && pin != REALTIME_PIN_ID &&
+	       cfg->dev_ids > 0);
```
this is IPC data, right? Wrong IPC data shouldn't crash the firmware. Alco `cfg` can be `NULL` here for removing the module, we shouldn't crash in that case

tobonex replied:
This mistake, and some of the other ones, comes from the fact that I tried to make the code as similar to ACE as I could, assuming that original logic was correct. Looking back, it really seems pretty inconsistent.

ACE used `debug_assert` here, and it seems it did actual assert when debug was false. Not sure what the idea was here. I will just change it to return.

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,190 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+	ret = unregister_modules_list(&priv_data->fmt,
+				      &priv_data->fmt_device_list.device_list_[pin],
+				      pin);
+	assert(ret == 0);
```
same here, just supplying the same pin number twice or similar would cause an error here?

lyakh reviewed (changes requested):
not a complete review, let's fix these issues first. Maybe this should be a draft PR for now

tobonex commented on src/ipc/ipc4/handler.c:
```diff
@@ -828,6 +828,99 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
+			const struct byte_array_simple array = {
+					(uint8_t *)&tlv->param_data[0], tlv->param_size };
+
+			ret = drv->ops.set_large_config(dev, tlv->param_id, init_block,
```
Good catch. From what I can see here, set_large_config implementations do not check for `dev == NULL` either , and first thing those do is trying to access  `dev->priv_data`.

The problem is, I based this function on ipc4_get_large_config_module_instance which does the same thing. Here we also write to dev variable only if module_id is not 0 (because base_fw has no dev apparently). 

After searching a bit I saw that base_fw also implements large config set in `basefw_set_large_config` function. It takes comp_dev argument just to be compatible with the interface, but assumes it is NULL as it is not used. 

So, it seems it really can be called with `dev == NULL` in this one case.

kv2019i commented on src/ipc/ipc4/handler.c:
```diff
@@ -828,6 +831,89 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
+	if (init_block) {
+		// for initial block use param_id from tlv, otherwise use built-in
+		// module instance handling mechanism (there is no default impl of
+		// caching large_param_id - @see LargeConfigSet)
```
C-style comments here as well

kv2019i commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,190 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+		if (fmt_device_list->modules_list_item_[module_slot_idx] == mi_ptr)
+			return -EINVAL; // was ADSP_INVALID_REQUEST;
+	}
+	//add module to first available field
```
C++ comment

kv2019i commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,190 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+static inline int alloc_fmt_module_list_item(struct kpb_fmt_dev_list *fmt_device_list,
+					     struct comp_dev *mi_ptr, struct comp_dev ***item)
+{
+	//check if module already added?
```
C++ comment

kv2019i reviewed (commented):
Mostly minor style issue. I understand this is based on code with C++ background, so some of the conversion artifacts can be omitted, but I did find at least one odd case of error handling.

As for functionality, I don't know kpb well enough to really do meaningful commentary and would welcome a review round by someone more familiar with kpb.

lgirdwood commented on src/include/ipc4/module.h:
```diff
@@ -51,6 +54,23 @@ enum sof_ipc4_module_type {
+struct byte_array_simple {
+	uint8_t *data;
+	uint32_t size;
+};
```
Is byte_array_simple part of the IPC ABI ?

kv2019i commented on src/audio/kpb.c:
```diff
@@ -2386,14 +2591,41 @@ static int kpb_set_large_config(struct comp_dev *dev, uint32_t param_id,
+
+		if (nr_of_modules != 0) {
+			cfg = (struct kpb_task_params *)array->data;
+			if (!cfg)
```
@tobonex how about this? I see you addressed rest of my concerns

tobonex replied:
I will take care of it next time I push

btian1 commented on src/include/sof/audio/kpb.h:
```diff
@@ -171,6 +171,7 @@ struct kpb_micselector_config {
 	/* channel bit set to 1 implies channel selection */
 	uint32_t mask;
 };
+
```
extra no need change

btian1 commented on src/include/ipc4/kpb.h:
```diff
@@ -19,4 +19,16 @@
+	/* Mic selector for client - sets microphone id for real time sink mic selector
+	 * IPC4-compatible ID - please do not change the number
+	 */
+	KP_BUF_CLIENT_MIC_SELECT = 11,
```
how about 101? left more extend for this?

tobonex replied:
It is defined as 11 in reference FW. There are some IPC commands in-between those two that are not defined in sof yet.

tobonex commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,192 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+static int ConfigureFastModeTask(struct comp_dev *kpb_dev, const struct kpb_task_params *cfg,
+				 size_t pin)
+{
+	if (pin >= KPB_MAX_SINK_CNT || pin == REALTIME_PIN_ID || cfg->dev_ids <= 0)
```
In reference FW, there was this check:
`cfg->dev_ids <= 0`
(checking if flexible array member is 0, or null as its value is a memory address in debug)
Can flexible array member be NULL at all? Removing this one.

lyakh replied:
in practice that condition will never be satisfied, and it doesn't really make sense. Currently `dev_ids` is at offset 4 inside `struct kpb_task_params` so, e.g. if `cfg == -4` then `cfg->dev_ids == 0`. If `cfg == -8` **and** the comparison is performed as `signed` then `cfg->dev_ids < 0`. But actually the compiler can even complain about that comparison between a pointer and an integer. You can only compare *compatible pointers* to each other. You can compare two `int *` pointers but IIRC you'll get a warning if you compare `int *` and `char *`

kv2019i commented:
@tobonex CI is failing on build, can you check:

https://github.com/thesofproject/sof/actions/runs/6151677351/job/16692249787?pr=8005
```
zep_workspace/sof/src/ipc/ipc4/handler.c: In function 'ipc4_set_large_config_module_instance':
/zep_workspace/sof/src/ipc/ipc4/handler.c:1032:67: error: passing argument 8 of 'ipc4_set_vendor_config_module_instance' discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
```

kv2019i reviewed (commented):
A few C++'ism, but I can live with those. The CI fail needs to be fixed before merge, and I'd appreciate some comment on why FMT is needed by KPB. Otherwise looks good to go, thanks @tobonex !

tobonex commented on src/ipc/ipc4/handler.c:
```diff
@@ -823,6 +826,79 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
+	}
+	ret = drv->ops.set_large_config(dev, param_id, init_block,
+		final_block, data_off_size, (uint8_t *)(&array));
+	return ret;
```
Fixed this, but ret is still used elsewhere.

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,22 +2398,202 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+#endif
+
+		if (!dev)
+			return -EINVAL;
```
so for IPC3 this function will return `-EINVAL`, can we just make that explicit somehow?

tobonex replied:
Yes, I do not support IPC3 here yet.  About initialisation of dev and the check - I could put those into the #ifdef, and if IPC4 is not defined just return in #else clause . Also I could add comment saying that  IPC3 support is not here yet. Does it sound good?

lyakh replied:
IPC3 isn't supported at all by this, right? So I'd really put something like
```
if (IS_ENABLED(CONFIG_IPC_MAJOR_3))
    return -EINVAL;
```
somewhere very early, into some preparation function without which the whole fast mode task won't work

tobonex replied:
Added this.
I'm also considering putting my new kpb.c functions in `#ifdef CONFIG_IPC_MAJOR_4` so that it's more explicit (instead of just this part).

lyakh replied:
> I'm also considering putting my new kpb.c functions in `#ifdef CONFIG_IPC_MAJOR_4` so that it's more explicit (instead of just this part).

I think yes, that would be the best for now. You do need a preprocessor check there because of calling an IPC4 specific function, so just using `IS_ENABLED(CONFIG_IPC_MAJOR_3)` unfortunately isn't enough - that wasn't a very good idea, sorry. So, yes, I think putting everything under `#if IPC4` would be the best option ATM

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,22 +2398,202 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+		if (ret < 0)
+			return ret;
+	}
+	return ret;
```
could do `return 0` here and remove initialisation above

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,22 +2398,202 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+		uint32_t outpin_id = extended_param_id.part.parameter_instance;
+
+		if (outpin_id >= KPB_MAX_SINK_CNT)
+			return -EINVAL;
```
this is checked twice - here and in the function. I'd remove this check.

tobonex commented:
Refactored some of the code so that it makes more sense:
- removed `struct byte_array_simple` that I previously added - managed without it
- removed some duplicated checks
- `configure_fast_mode_task` function no longer accepts null in cfg argument as it made error checking messy

tobonex commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,192 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+	dev = ipc4_get_comp_dev(comp_id);
+#endif
+	return dev;
+}
```
Yeah, used it only once anyway, removed it.

tobonex commented on src/ipc/ipc4/handler.c:
```diff
@@ -823,6 +826,79 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
+			}
+			/* Align up to a multiple of 4 */
+			tlv = (struct sof_tlv *)((const uint8_t *)tlv +
+					TLV_DATA_OFFSET + ALIGN_UP(tlv->length, 4));
```
this `tlv_next` looks convenient, but...
is does not have the `ALIGN_UP`. Reference FW aligned it up, so I guess that we also need it here. I do not know all of the possible IPC cases so I'm not sure if it's actually necessary or not.

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,22 +2398,202 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+	clear_fmt_modules_list(&priv_data->fmt_device_list, pin);
+
+	/* When modules count IS 0 we only need to remove modules from Fast Mode. */
+	if (cfg->number_of_modules > 0) {
```
the function itself also has checks for `cfg->number_of_modules > 0`  and does nothing if that isn't satisfied. I'm remove this and other already known to hold checks from the function - see comment there

tobonex replied:
Left this one, removed other ones.

lyakh commented on src/audio/kpb.c:
```diff
@@ -2378,22 +2398,202 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+	if (!modules_to_prepare)
+		return -EINVAL;
+	if (outpin_idx >= KPB_MAX_SINK_CNT)
+		return -EINVAL;
```
All of the above checks and the below `modules_to_prepare->number_of_modules != 0` are checked by the caller. I think they can be removed here, or you can turn them into assert()s with a comment

lyakh commented on src/ipc/ipc4/handler.c:
```diff
@@ -823,6 +826,79 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
+						  bool init_block,
+						  bool final_block,
+						  uint32_t data_off_size,
+						  const char *data,
```
I'd say we have mostly 3 options for this:
1. `void *` - because functions like `write(2)` use it, but arithmetics is difficult with it
2. `char *` - because `char` is simple, everybody knows, that it's 1 byte...
3. `uint8_t *` - because it's explicitly a pointer to a byte array

SOF uses all 3 options in different cases. In general I'd probably use (1), but since we already use them all... Feel free to make your choice, my comment was a proposal, not a hard requirement.

tobonex replied:
Leaving this one as it is for now.

tobonex commented:
> A few C++'ism, but I can live with those. The CI fail needs to be fixed before merge, and I'd appreciate some comment on why FMT is needed by KPB. Otherwise looks good to go, thanks @tobonex !

Originally, in case if WoV event, KPB used FMT to drain history buffer and send this data to Host ASAP. Normal tasks were too slow. We still need it for the same reason.

tobonex commented on src/audio/kpb.c:
```diff
@@ -2378,6 +2399,192 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+				      new_list_item_ptr);
+		if (ret < 0)
+			return ret;
+	}
```
I didn't read that much into the original FMT functionality code yet, but I do have some info about it though. In reference FW, executing the FMT also triggered processing of some LL modules whenever there are available DSP cycles. I'm guessing those modules are the ones provided in the IPC here.

I agree that the way the lists were (and are) handled was kind of confusing, but I'm not sure yet what can I refactor here. It's either some important logic, or just some C++'ism.

The idea for now was to implement it up to this point so that the tests related to the FMT config IPC pass. After FMT functionality actually gets implemented it will be easier to see if we can simplify the config.

kv2019i commented:
@tobonex wrote:
> Originally, in case if WoV event, KPB used FMT to drain history buffer and send this data to Host ASAP. Normal tasks were too slow. We still need it for the same reason.

Ack, thanks, this is a sufficient description.

lyakh commented on src/include/sof/audio/kpb.h:
```diff
@@ -159,18 +163,88 @@ struct history_data {
+ *		  |
+ *		fmt.device_list(device_list*)
+ *
+ *
```
2 empty lines above and below the description should be removed IMHO

lyakh commented on src/include/sof/audio/kpb.h:
```diff
@@ -159,18 +163,88 @@ struct history_data {
+ *		dev_list.device_list(comp_dev**)
+ *		  ^
+ *		  |
+ *		fmt.device_list(device_list*)
```
not sure what this structure means. Are these elements of the `struct device_list::devs[]` array above? Maybe we can find some easier (at least for me) representation. If this is an array, maybe arrange entries in lines
```
array[]
    +-------> element 1
    +-------> element 2
    +...
    +-------> element N
```
or something similar. The above looks to me like things pointing to each other, or calling each other, or similar

lyakh reviewed (approved):
comments have been addressed, thanks! Haven't checked what exactly changed from previous versions, but at least locations, that I commented on, look better now
